File: src/main/java/com/anatoliapark/nursinghome/service/base/EntityService.java
package com.anatoliapark.nursinghome.service.base;

import com.anatoliapark.nursinghome.entity.base.BaseConstantEntity;
import com.anatoliapark.nursinghome.entity.base.BaseEntity;
import com.anatoliapark.nursinghome.repository.base.EntityRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.ExampleMatcher;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class EntityService extends BaseService{
    @Autowired
    protected EntityRepository entityRepository;

    public <T extends BaseEntity> T save(BaseConstantEntity constantEntity) {
        if(constantEntity != null)
            return (T)entityRepository.save(constantEntity);
        return null;
    }

    public <T extends BaseEntity> Optional<T> find(Long id, Class<T> c) {
        ExampleMatcher matcher = ExampleMatcher.matchingAll();
        T instance = createInstance(c);
        instance.setId(id);
        if(id != null)
            return entityRepository.findOne(Example.of(instance, matcher));
        return null;
    }

    public <T extends BaseEntity> List<T> findAll(Class<T> c) {
        T instance = createInstance(c);
        return entityRepository.findAll(Example.of(instance));
    }

    public <T extends BaseEntity> void delete(Long id, Class<T> c){
        T instance = createInstance(c);
        instance.setId(id);
        entityRepository.delete(instance);
    }
}

File: Inventory/src/main/kotlin/com/experiment/inventory/service/base/EntityService.kt
package com.experiment.inventory.service.base

import com.experiment.inventory.entity.base.BaseEntity
import com.experiment.inventory.repository.EntityRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Example
import org.springframework.data.domain.ExampleMatcher
import org.springframework.stereotype.Service
import java.util.*
import kotlin.reflect.KClass

@Service
class EntityService<T>:BaseService() where T: BaseEntity {

    @Autowired
    private var entityRepository: EntityRepository<T>? = null

    fun save(baseEntity: T): T? {
        return entityRepository!!.save(baseEntity)
    }

    fun find(id: Long, c: KClass<T>): Optional<T> {
        val matcher = ExampleMatcher.matchingAll()
        val instance: T = createInstance(c)
        instance.id=id
        return entityRepository!!.findOne(Example.of(instance, matcher))
    }

    fun findAll(c: KClass<T>): List<T> {
        val instance: T = createInstance(c)
        return entityRepository!!.findAll(Example.of(instance))
    }

    fun delete(id: Long, c: KClass<T>) {
        val instance: T = createInstance(c)
        instance.id = id
        entityRepository!!.delete(instance)
    }
}